Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+      let decoder = GRPCMessageDeframer(
+        maximumPayloadSize: previousState.maximumPayloadSize,
+        decompressor: self.decompressor
+      )
+      self.deframer = NIOSingleStepByteToMessageProcessor(decoder)
+      
+      self.inboundMessageBuffer = .init()
+    }
+  }
+  
+  struct ClientOpenServerOpenState {
+    var framer: GRPCMessageFramer
+    var compressor: Zlib.Compressor?
+    
+    let deframer: NIOSingleStepByteToMessageProcessor<GRPCMessageDeframer>
+    var decompressor: Zlib.Decompressor?
+    
+    var inboundMessageBuffer: OneOrManyQueue<[UInt8]>
+    
+    init(previousState: ClientOpenServerIdleState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.deframer = previousState.deframer
+      self.decompressor = previousState.decompressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+    }
+  }
+  
+  struct ClientOpenServerClosedState {
+    var framer: GRPCMessageFramer
+    var compressor: Zlib.Compressor?
+    
+    let deframer: NIOSingleStepByteToMessageProcessor<GRPCMessageDeframer>
+    var decompressor: Zlib.Decompressor?
+    
+    var inboundMessageBuffer: OneOrManyQueue<[UInt8]>
+    
+    init(previousState: ClientOpenServerOpenState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.deframer = previousState.deframer
+      self.decompressor = previousState.decompressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+    }
+  }
+  
+  struct ClientClosedServerIdleState {
+    var framer: GRPCMessageFramer
+    var compressor: Zlib.Compressor?
+    
+    let deframer: NIOSingleStepByteToMessageProcessor<GRPCMessageDeframer>
+    var decompressor: Zlib.Decompressor?
+    
+    var inboundMessageBuffer: OneOrManyQueue<[UInt8]>
+    
+    init(previousState: ClientOpenServerIdleState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.deframer = previousState.deframer
+      self.decompressor = previousState.decompressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+    }
+  }
+  
+  struct ClientClosedServerOpenState {
+    var framer: GRPCMessageFramer
+    var compressor: Zlib.Compressor?
+    
+    let deframer: NIOSingleStepByteToMessageProcessor<GRPCMessageDeframer>
+    var decompressor: Zlib.Decompressor?
+    
+    var inboundMessageBuffer: OneOrManyQueue<[UInt8]>
+    
+    init(previousState: ClientOpenServerOpenState) {
+      self.framer = previousState.framer
+      self.compressor = previousState.compressor
+      self.deframer = previousState.deframer
+      self.decompressor = previousState.decompressor
+      self.inboundMessageBuffer = previousState.inboundMessageBuffer
+    }
+  }
+}
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+struct GRPCStreamStateMachine {
+  private var _stateMachine: GRPCStreamStateMachineProtocol
```

Comment:
We should avoid going through a protocol here and just have an enum of the two different state machines.

Reply:
I thought having the code completely separate instead of having each method be a big `switch` case and containing the logic for both sides was slightly more readable. But yes probably the performance cost is too high and I understand why we'd want to do it the other way. I can change it.